gustaf/ffd.py

```diff
@@ -1,4 +1,455 @@
+        -------
+        MESH_TYPES
+            Current Mesh with the deformation according to the current spline.
+        """
+        self._deform()
+        return self._mesh.copy()
+
+    @mesh.setter
+    def mesh(self, mesh: MESH_TYPES):
+        """
+        Sets mesh. If it is first time, the copy of it will be saved as 
+        original mesh. If spline is already defined and in transformed status, 
+        it applies transformation directly.
+
+        Parameters
+        -----------
+        mesh: MESH_TYPES
+            Mesh used for the FFD
+
+        Returns
+        --------
+        None
+        """
+        if self._spline is None:
+            # Define a default spline if mesh is given but no spline
+            par_dim = mesh.vertices.shape[1]
+            self.spline = with_bounds(
+                [[0]*par_dim, [1]*par_dim],
+                mesh.bounds)
+
+        self._logi("Setting mesh.")
+        self._logi("Mesh Info:")
+        self._logi(
+            "  Vertices: {v}.".format(v=mesh.vertices.shape)
+        )
+        self._logi(
+            "  Bounds: {b}.".format(b=mesh.get_bounds())
+        )
+        self._mesh = mesh.copy()  # copy to make sure given mesh stays untouched
+        self._check_dimensions()
+        self._scale_mesh_vertices()
+        self._is_calculated = False
+
+    @property
+    def spline(self):
+        """
+        Returns a copy of the spline. Please use the setter to explicitly make
+        changes to the spline.
+
+        Parameters
+        -----------
+        None
+
+        Returns
+        --------
+        self._spline: Spline 
+        """
+        self._logd("Returning copy of current spline.")
+        return self._spline.copy() if self._spline is not None else None
+
+    @spline.setter
+    def spline(self, spline: SPLINE_TYPES):
+        """
+        Sets spline. The spline parametric range bounds will be converted 
+        into the bounds [0,1]^para_dim.
+
+        Parameters
+        -----------
+        spline: SPLINE_TYPES
+            New Spline for the next deformation
+
+        Returns
+        --------
+        None
+        """
+        self._spline = spline.copy()
+        self._scale_parametric_dimension_to_hypercube()
+
+        self._logi("Setting Spline.")
+        self._logi("Spline Info:")
+        self._logi(
+            f"  Parametric dimensions: {spline.para_dim}."
+        )
+        self._is_calculated = False
+
+    def _scale_parametric_dimension_to_hypercube(self):
+        """Scales all knot_vectors of the spline to a range of [0,1]
+        """
+        if not self._is_parametric_room_hypercube(self._spline):
+            knv = []
+            for knot_vector in self._spline.knot_vectors:
+                knot_vector = [kv_v - knot_vector[0] for kv_v in knot_vector]
+                knot_vector = [kv_v / knot_vector[-1] for kv_v in knot_vector]
+                knv.append(knot_vector)
+            self._spline.knot_vectors = knv
```

Comment:
Hey, just wanted to let you know that there is a function in `splinepy.SplineBase`, from which splines inherit, called `normalize_knot_vectors`, that you could call here in order to reuse existing code